### PR TITLE
Refactor: Centralize file operations in common_utils

### PR DESCRIPTION
**Pull request by Raqeeb27**
### Overview

This pull request refactors the hiding and recovery functionality to utilize functions from the `common_utils` module.

### Changes

* **common\_utils.py:**
    * Added `move_file()` and `validate_ext()` methods.
* **hiding.py & recovery.py:**
    * Utilized the methods from `common_utils.py` for file moving and extension validation operations.

### Purpose

* **Reduce code duplication:** The file moving and extension validation logic was present in both the `hiding.py` and `recovery.py` files. This change centralizes that logic in `common_utils.py`.

### Testing

* Verified proper integration of `common_utils` functions within the hiding and recovery processes.
* Confirmed that file hiding and recovery operations function as expected after the refactor.

### Notes

Please review these changes and provide feedback before merging.

### CODE REVIEW

File: modules/common_utils.py [hunk 1]
```diff
@@ -98,3 +99,71 @@ def ext_to_app_path(ext: str, app_path_db: dict[str, dict[str, list | str]]) ->
...
+    This function attempts to move a file using `os.rename` for efficiency. 
+    If `os.rename` fails due to a `PermissionError` or other issues, it falls 
+    back to using `shutil.move` to complete the operation.
+
+    Args:
+        src_path (str): The source file path.
+        dest_path (str): The destination file path.
+
+    Returns:
+        bool: True if the file was successfully moved, False otherwise.
```
Comment (hanXen) [inline modules/common_utils.py:117]: The type annotation and the actual code indicate that the function returns a `bool | None` type. However, the docstring states that it returns only a `bool` type. In my opinion, explicitly returning `False` and aligning the implementation with the docstring to always return a `bool` type would make the code more intuitive.

File: hiding.py [hunk 2]
```diff
@@ -229,6 +213,9 @@ def main(file_path: str = "", is_test: bool = False) -> None:
             # "help" is hardcoded as the hidden_dir_key for testing purposes
             target_list.append(make_shortcut(file_path, ext_icon_dict, hidden_dir_key="help"))
     else:
+        file = validate_extension(file_path)
+        if not file:
+            return
```
Comment (hanXen) [inline hiding.py:218]: How about simplifying this by removing the unnecessary `file` variable, since it's not used later?

``` python
if not validate_extension(file_path):
    return
```

File: modules/common_utils.py [hunk 3]
```diff
@@ -98,3 +99,71 @@ def ext_to_app_path(ext: str, app_path_db: dict[str, dict[str, list | str]]) ->
...
+        except OSError as e:
+            print(e)
+
+
+def validate_extension(file_path: str) -> str | None:
+    """
+    Validates the file extension of a given file path.
+
+    This function ensures that the file has a valid extension. If the extension
+    is missing or invalid, it prints an error message and exits the program.
```
Comment (hanXen) [inline modules/common_utils.py:154]: The function doesn't exit the program; it returns `None` if the extension is invalid or missing, and the program termination occurs elsewhere. I think it would be better to change "exits" to "None" in the docstring to reflect the actual behavior.

File: modules/common_utils.py [hunk 4]
```diff
@@ -98,3 +99,71 @@ def ext_to_app_path(ext: str, app_path_db: dict[str, dict[str, list | str]]) ->
...
+            print("[!] Please close the file and try again.\n")
+    except OSError:
+        try:
+            shutil.move(src_path, dest_path)
+            return True
+        except OSError as e:
+            print(e)
+
+
+def validate_extension(file_path: str) -> str | None:
```
Comment (hanXen) [inline modules/common_utils.py:149]: There are several points to discuss regarding this function:

1. **Role of `validate_extension()`**

   * As I understand it, this function returns `None` if there is no file extension, and if an extension exists, it removes whitespace from the `file_path` and returns it.
   * If my understanding is correct, the function name does not align with its behavior. The name `validate_extension()` suggests that it includes logic to comprehensively validate extensions.
   * In this project, the compatible extensions are clearly defined, and in fact, the `make_shortcut()` function in `hiding.py` contains logic to validate extensions.
   * In conclusion, the role of this function seems unclear. I tried to think of a suitable function name but found it challenging.
   * Would it make sense to split its functionality into the following parts?

     * Remove leading and trailing whitespace from the `file_path` (as Windows ignores whitespace around file names).
     * Parse the file extension.
     * Check if the file has no extension or an unsupported extension.

2. **Bugs, Questions, and Suggestions in the Function**

   * The current code does not filter out file names ending with `.` such as `test.`.
   * The current code incorrectly returns `test.` for file names like `test.  txt`.

     * Is there a need for a complex return like `return file_path[:file_path.rfind(ext) + len(ext)]`?
     * A simple `return file_path` is sufficient to remove unnecessary whitespace.
     * Invalid extensions like `.  txt` can be filtered out later since they are unsupported.

3. **Summary**
   If my understanding of the function's role is correct, it would be better to organize it as follows:

    * Remove leading and trailing whitespace from the `file_path`.
    * Parse the file extension & check if the file has no extension or an unsupported extension (and decide whether this should be placed in `common_utils` or handled exclusively in `make_shortcut()`).

**Review by hanXen — COMMENTED** [review]
The creation of the `move_file()` function is an excellent step toward reducing code duplication and achieving centralized management. I've provided a comprehensive review for the `validate_extension()` function, along with several minor comments. Thanks for your contribution.

**Comment by Raqeeb27** [comment]
Thank you again for your thorough review and for highlighting the areas for improvement in the recent changes. I truly appreciate you taking the time to provide such detailed feedback.

I sincerely apologize for the oversight regarding the function documentation. I failed to update the docstrings after testing and finalizing the code modifications. I will ensure this is rectified immediately.

Furthermore, I acknowledge the mistake of overlooking the `make_shortcut()` function in `hiding.py`, the usage of `rstrip()` for extension validation. I will definitely update that part of the code.

* Regarding the naming of `validate_extension()`, I agree that its current name doesn't accurately reflect its functionality. While its current implementation isn't strictly required during the hiding process (as only supported extensions are considered there), it plays a crucial role in the recovery process by ensuring the correct shortcut path/name is used to recover the original file.

* Regarding the whitespace in filenames, while Windows may ignore whitespace in some scenarios, it does not consistently do so, particularly when renaming files via the command line (leading whitespaces). My intention was to handle these edge cases proactively to prevent potential issues and ensure the code's robustness for such cases.

* You also pointed out the issue with filenames ending in a dot (e.g., `test. `) and invalid extensions like `. txt`. Thank you for bringing this to my attention. I will implement the suggested check `if not ext or ext.strip() == ".":` to address this.

* Finally, regarding the complex return statement `return file_path[:file_path.rfind(ext) + len(ext)]`, my intention was to handle potential suffixes appended to filenames, such as "- Copy" or "- Kopie," or any other similar additions. The goal was to ensure that the returned path included the extension but excluded any such suffixes. To improve readability, I agree that splitting the return statement would be beneficial:

```python
end_of_filename = file_path.rfind(ext) + len(ext)
return file_path[:end_of_filename]
```

I support your suggestion to split the functionality of `validate_extension()`, and I think `process_filepath_for_extension()` would be a better name to reflect its functionality. However, I agree that finding the perfect name for this function remains a bit challenging. Furthermore, I think it would be beneficial to place this function in `common_utils` and implement it in both the hiding and recovery processes for more robust filename sanitization.

Here is the updated function:
```python
def process_filepath_for_extension(file_path):
    directory = os.path.dirname(file_path)
    file_path = os.path.join(directory, os.path.basename(file_path).strip())
    ext = os.path.splitext(file_path)[1].split(" ")[0]
    if not ext or ext.strip() == ".":
        print(f"[!] Error in filename {file_path}: Invalid or Missing file extension.")
        return None
    else:
        end_of_filename = file_path.rfind(ext) + len(ext)
        return file_path[:end_of_filename]
```

**Thank you again for your valuable feedback.** Please review this response for updating the pull request accordingly.

**Comment by hanXen** [comment]
Thank you for your detailed response. Based on my understanding so far, the `process_filepath_for_extension()` function seems to aim to include the following functionalities:

1. Remove unnecessary leading and trailing whitespaces from the file path.
2. Validate the file name format (including checks for files without extensions).
3. Remove suffixes like `- Copy` (I didn’t catch this intent during the previous reviews).

I have a few questions and suggestions regarding the `process_filepath_for_extension()` function to clarify and ensure alignment. Please let me know if there are any misunderstandings on my part.

1. Regarding point 3, does this function actually remove suffixes like `- Copy`?

   * If so, when I tested it, the suffix didn’t appear to be removed. Could you clarify how the suffix removal works?
   * Based on the current implementation, copying an `.lnk` file doesn’t seem to copy the original file alongside it. What benefit does removing suffixes like `- Copy` provide?
   * If a copied `.lnk` file is recovered, the original `.lnk` file remains intact. Is this the intended behavior?
     * As the original `.lnk` file remains intact when a copied `.lnk` file is recovered, this functionality seems incomplete and requires further refinement.
     * Furthermore, if we eventually implement the functionality to copy the original file along with the `.lnk` file, it may conflict with the functionality being implemented now.
   * Perhaps we could consider disallowing the copying of .lnk files temporarily until a more robust implementation is in place. What are your thoughts?

2. Regarding point 2, the `make_shortcut()` function in `hiding.py` already performs extension validation.

   * Would it make sense for `recovery.py` to rely on the validation already done in `hiding.py`? If this function isn’t meant to remove suffixes like `- Copy`, is there any other issue being addressed here?

3. Regarding point 1, Windows handles leading and trailing whitespaces in filenames without any issues.

   * If the filenames are used as provided by the user, wouldn’t it suffice to leave them as is? Are there any specific cases where handling these whitespaces proactively would be necessary?
     * If we’re validating only against a whitelist of supported extensions, wouldn’t that address any potential issues?
   * In conclusion, if functionalities 2 and 3 are not necessary, is functionality 1 also unnecessary?

Thank you again for your thorough response and for your diligent work on this feature. I look forward to hearing your thoughts and aligning on the next steps.

**Comment by Raqeeb27** [comment]
Regarding the scenario where a user wants to copy a hidden file: I agree that the current functionality of the project, which would involve recovering the hidden file, making a copy, and then re-hiding both, is not a feasible approach for the user. Therefore, my initial thought was that the user could copy the link file itself and then use the copied link file.

However, I agree with the significant issue that both link files would point to the same hidden file. Ideally, a future feature should be implemented to copy the hidden file as well and update the hash value of the copied link file to correspond with the newly copied hidden file.

I also agree that based on the current functionality, recovering a copied link file leaves the original link file intact. The issue you raised about copying the hidden file and updating the copied link file's hash would indeed resolve this.

For the time being, if we consider a scenario where the copied link file **does** point to a newly copied hidden file, there are issues with the naming convention of the copied link file (specifically, the "- Copy" suffix).

* Normally, when a regular file (e.g., `test.txt`) is copied within the same folder using Ctrl+C and Ctrl+V, the copied file is named `test - Copy.txt` (the "- Copy" suffix is appended before the file extension).
* However, when a link file (e.g., `test.txt.lnk`) is copied within the same folder, the copied link file is named `test.txt - Copy` (the "- Copy" suffix is appended after the file extension).

This creates the following problems:

* Upon recovery of the hidden file linked to the copied shortcut, the hidden file is also incorrectly renamed to `text.txt - Copy`, which appears to corrupt the file extension. This was the reason behind my attempt to remove the "- Copy" suffix.
* However, removing the "- Copy" suffix introduces another issue when trying to remove the link file:
    * The line of code in `recovery.py`: `shortcut_path = f"{original_name}{original_ext}.lnk"` attempts to locate the link file, but the absence of the "- Copy" suffix in the filename prevents the code from correctly determining the link file's existence.

Considering these issues related to the copying scenario and our suggestion for a more comprehensive feature, I now prefer to drop this consideration of copying link files until we can implement a feature that also copies the hidden files and updates the hash value of the copied link file accordingly.

Given that we are setting aside the copying scenario for now, I also agree that there is no immediate need to address points 1 and 2 from the previous discussion.

In conclusion, I believe that for the time being, the `process_filename_for_extension()` function is not necessary, as the copying scenario that prompted its consideration will not be implemented.

**Thank you for the very detailed review of the issue**, which has greatly clarified the functionality and direction of the code for me.

**Comment by Raqeeb27** [comment]
Hello there,

How about adding an option like 'Create a copy of this link file' to the right-click menu for copying files?

I've created a new file, `copy_link_file.py`, which performs the following actions:

* Extracts the hash value from the link file and finds it in the `hash_table`.
* Creates a copy of the corresponding hidden file within the same directory, giving it a random filename.
* Generates the hash of the copied hidden file and creates a new link file. This new link file will have its target path updated with the new hash value.
* Adds an entry to both the `mapping_table` and the `hash_table` for the newly created hidden file path and its associated link file path.

This new functionality aims to efficiently create copies of both the link file and its corresponding hidden file. I've tested the script, and it works seamlessly as intended.

However, this script was just for testing purposes, and we can integrate this functionality into `hiding.py` itself by introducing a `--copy_link_file` argument.

Please let me know your thoughts on this approach.

**Comment by hanXen** [comment]
This is a great approach. Instead of copying the hidden file within the same directory, it might be better to randomly select a hidden directory for the copy. This would enhance randomness while maintaining functionality without introducing any issues.

Regarding the proposed feature, here are some potential concerns that come to mind:

* Users habitually using CTRL+C and CTRL+V instead of the right-click menu

  * File copying is a common and instinctive action, and we cannot ignore user habits.
  * We might need to explore ways to guide users to use the right-click menu for this functionality.
  * Currently, I don’t have a clear solution for this. One possible direction might be leveraging Windows API hooking or similar techniques.

I believe this issue is not something we can resolve at this stage. For now, I suggest implementing the feature and revisiting this concern later.

As for next steps, how about closing PR #16 for now and submitting separate PRs: one for the `move_file()` feature and another for the `lnk` file copy functionality? (The `validate_extension()` function no longer seems necessary.)
Alternatively, if it’s more convenient, you could modify PR #16 to include only the `move_file()` feature. Please proceed in whichever way works best for you.

Thank you for sharing your excellent ideas, and have a great day!

**Comment by hanXen** [comment]
This is a great approach. Instead of copying the hidden file within the same directory, it might be better to randomly select a hidden directory for the copy. This would improve randomness while maintaining functionality.

Regarding the proposed feature, here are some potential concerns that come to mind:

* Users habitually using CTRL+C and CTRL+V instead of the right-click menu

  * File copying is a common and instinctive action, and we cannot ignore user habits.
  * We might need to explore ways to guide users to use the right-click menu for this functionality.
  * Currently, I don’t have a clear solution for this. One possible approach could involve leveraging Windows API hooking or similar techniques.

I believe this issue is not something we can resolve at this stage. For now, I suggest implementing the feature and revisiting this concern later.

As for next steps, how about closing PR #16 for now and submitting separate PRs: one for the `move_file()` feature and another for the `lnk` file copy functionality? (The `validate_extension()` function no longer seems necessary.) Alternatively, if it’s more convenient, you could modify PR #16 to include only the `move_file()` feature. Please proceed in whichever way works best for you.

In line with collaborative practices, for minor changes, your current approach works well. However, for more significant changes, it would be helpful to document the proposed improvement in the **Issues** tab. This allows for a clear record of the discussion and rationale, making it easier to track progress and reference in future PRs. If possible, consider linking the PR to the relevant issue to ensure better traceability and provide context for the changes.

Thank you for sharing your excellent ideas, and have a great day!

**Comment by Raqeeb27** [comment]
Thank you for your detailed feedback and the clear next steps. I appreciate your guidance on managing the pull requests and the suggestion regarding the Issues tab.

I agree that creating separate PRs would be a cleaner approach. Therefore, I prefer to close this current PR (#16) and will proceed to open a new pull request specifically for the `move_file()` functionality.

Thank you again for your valuable insights!
